Introduce the Dummy.

dixpac opened the pull request:
First step:
* Convert controller tests to integration tests
* Move storage configurations to dummy
* Remove unnecessary requires

r? @dhh 
We could slim down a bit this dummy, to remove rails components that are not used at the moment, ex: ActionCable ?

dixpac commented on test/test_helper.rb:
```diff
@@ -14,16 +12,14 @@
 require "yaml"
 SERVICE_CONFIGURATIONS = begin
-  YAML.load_file(File.expand_path("../service/configurations.yml", __FILE__)).deep_symbolize_keys
+  YAML.load_file(File.expand_path("../dummy/config/storage_services.yml", __FILE__)).deep_symbolize_keys
```
I've moved the configurations inside the dummy, otherwise dummy app will fail on the boot

dhh commented:
I'd welcome slimming this down. There are a ton of files that aren't relevant. But if you could otherwise bring in line with master, we can merge. Definitely would like to switch to integration tests.

dixpac commented on test/controllers/disk_controller_test.rb:
```diff
@@ -1,24 +1,31 @@
-    assert_equal "inline; filename=\"#{@blob.filename}\"", @response.headers["Content-Disposition"]
+    get rails_disk_blob_url(
+      filename: "hello.txt",
+      content_type: @blob.content_type,
```
Here I'd to pas content-length in order for tests to pass. @dhh  I've noticed that you changed the approach [Here](https://github.com/rails/activestorage/commit/5944850bc1259ca42381ce83d155ddd914b968c6#diff-a8a2fcba8c0f485c1902e82082c3d582R15).

Not sure if your intention was to pass content_type of to figure him out automatically ?

dhh replied:
Plan is to pass it in to the services, since they don't store it, and the blob has it.

dixpac commented on test/models/blob_test.rb:
```diff
@@ -35,6 +35,7 @@ class ActiveStorage::BlobTest < ActiveSupport::TestCase
   private
     def expected_url_for(blob, disposition: :inline)
-      "/rails/active_storage/disk/#{ActiveStorage.verifier.generate(blob.key, expires_in: 5.minutes, purpose: :blob_key)}/#{blob.filename}?disposition=#{disposition}&content_type=#{blob.content_type}"
+      query_string = { content_type: blob.content_type, disposition: disposition }.to_param
```
Rails encodes the url so I'h to encode the query_string in order to match it sucessfully

dixpac commented on test/service/disk_service_test.rb:
```diff
@@ -6,7 +6,7 @@ class ActiveStorage::Service::DiskServiceTest < ActiveSupport::TestCase
 
   test "url generation" do
-    assert_match /rails\/active_storage\/disk\/.*\/avatar\.png\?disposition=inline/,
+    assert_match /rails\/active_storage\/disk\/.*\/avatar\.png\?.+disposition=inline/,
```
Changed a regex a bit because of the url encoding

dixpac commented:
@dhh  I rebase it from the latest. Commit is a bit big(sorry) so I've added few comments on what I've changed from the first PR.

re: In the next commits/days I will try to slim this down

dixpac commented:
@dhh I slimmed this quiet a bit (maybe too much) by removing all rails components we are not using at the moment.


There is one more approach I've tried and that is to slim all the way to the smallest app possible, take a look [commit](https://github.com/dixpac/activestorage/commit/14e8fd80718b6e156330fa4c152cf1d3d5ed9e60) and here [small app](https://github.com/dixpac/activestorage/commit/14e8fd80718b6e156330fa4c152cf1d3d5ed9e60#diff-5d4a3b08ef4b09c90b866db12b8fc04fR8)
(this has just a few files)

What approach do you think is better ?

dhh commented on test/controllers/disk_controller_test.rb:
```diff
@@ -1,24 +1,31 @@
   test "sending blob as attachment" do
-    get :show, params: { filename: @blob.filename, encoded_key: ActiveStorage.verifier.generate(@blob.key, expires_in: 5.minutes, purpose: :blob_key), disposition: :attachment }
-    assert_equal "attachment; filename=\"#{@blob.filename}\"", @response.headers["Content-Disposition"]
+    get rails_disk_blob_url(
```
Reminds me that now this should be `rails_disk_service_url` since the DiskController no longer deals with blobs, but only with keys.

dhh commented:
I think it's good with the smallest possible app. We're basically just setting all this stuff up to be able to use the engine initialization and integration tests properly.

dhh commented:
(Needs some rebasing after George merged the direct upload feature for the disk controller).

dixpac commented:
@dhh  I've rebased and added small dummy app

kaspth commented:
Won't we just need to add back Action View and Sprockets when the direct upload JS arrives and has to be integration/system tested?

dixpac commented:
@kaspth yes we would! That was my initial approach just to remove action cable and action mailer and leave other components  in there  (for the future). But, there were was concerned with the size of the dummy so I just created it as small as "possible", to cover current use.

The full dummy app would be most flexible/easy for the future use of testing.

dhh commented:
Would probably be good to get AV and Sprockets back in, yeah, because
@javan is just about ready with his direct upload JS PR.

Thanks for working on this, Dino!

On Wed, Jul 26, 2017 at 12:48 PM, Dino Maric <notifications@github.com>
wrote:

> @kaspth <https://github.com/kaspth> yes we would! That was my initial
> approach just to remove action cable and action mailer and leave other
> components in there (for the future). But, there were was concerned with
> the size of the dummy so I just created it as small as "possible", to cover
> current use.
>
> The full dummy app would be most flexible/easy for the future use of
> testing.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rails/activestorage/pull/70#issuecomment-318130529>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAKtYsOuj7dhZuQ0qQI5AugC__NaXaXks5sR3vqgaJpZM4Ogb-S>
> .
>

dixpac commented:
No problem! Will get them back :)

On Jul 26, 2017 8:43 PM, "David Heinemeier Hansson" <
notifications@github.com> wrote:

> Would probably be good to get AV and Sprockets back in, yeah, because
> @javan is just about ready with his direct upload JS PR.
>
> Thanks for working on this, Dino!
>
> On Wed, Jul 26, 2017 at 12:48 PM, Dino Maric <notifications@github.com>
> wrote:
>
> > @kaspth <https://github.com/kaspth> yes we would! That was my initial
> > approach just to remove action cable and action mailer and leave other
> > components in there (for the future). But, there were was concerned with
> > the size of the dummy so I just created it as small as "possible", to
> cover
> > current use.
> >
> > The full dummy app would be most flexible/easy for the future use of
> > testing.
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/rails/activestorage/pull/70#issuecomment-318130529>,
> > or mute the thread
> > <https://github.com/notifications/unsubscribe-auth/
> AAAKtYsOuj7dhZuQ0qQI5AugC__NaXaXks5sR3vqgaJpZM4Ogb-S>
> > .
> >
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rails/activestorage/pull/70#issuecomment-318145730>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHFVJe-8EB1mhaAIcUPWk6tB7wPgZ3Yiks5sR4jkgaJpZM4Ogb-S>
> .
>

dhh commented:
There are a couple of conflicts against master, btw.

dixpac commented:
I've add back ActionView, Sprockets.... Now the dummy is larger because there is not we can remove I've removed ActionCable, ActionMailer and few config files.

dhh commented:
The build is currently failing. Can you have a look?

dixpac commented:
@dhh  oh, sorry I fixed it :)
